File: frontend/src/api/solutionLibrary.ts
import request from '@/utils/request'

export interface SolutionLibrary {
  id: number
  name: string
  description: string
  category: string
  version: string
  author: string
  icon: string
  pack_data: string        // OrbitPack JSON (非空=未应用, 空=已应用)
  stage_ids: string     // JSON string like "[1,2,3]"
  variable_ids: string
  hook_ids: string
  template_ids: string
  file_ids: string
  workflow_ids: string
  stage_count: number
  variable_count: number
  hook_count: number
  template_count: number
  file_count: number
  workflow_count: number
  created_at: string
  updated_at: string
}

export interface OrbitPack {
  apiVersion: string
  kind: string
  metadata: {
    name: string
    description: string
    category: string
    version: string
    author: string
  }
  materials?: Array<{
    name: string
    size: number
    md5: string
    downloadUrl?: string
  }>
  globalVariables?: Array<{
    key: string
    type: string
    value: string
    description: string
    group: string
  }>
  hooks?: Array<{
    name: string
    description?: string
    module: string
    params?: string
    timeout?: number
    ignore_errors?: boolean
    retries?: number
    delay?: number
  }>
  workflowTemplates?: Array<{
    name: string
    description?: string
    content: string
  }>
  stages?: Array<{
    name: string
    description?: string
    machineGroup?: string
    tasks: Array<{
      ref: number
      name: string
      module: string
      order: number
      params?: string
      when?: string
      hooks?: string
      loop?: string
      timeout?: number
      ignore_errors?: boolean
      retries?: number
      delay?: number
      register?: string
    }>
  }>
  machineGroups?: Array<{
    name: string
    description?: string
  }>
  workflows?: Array<{
    name: string
    description?: string
    stageGroups?: Array<{
      name: string
      description?: string
      order: number
      mode: string
      stages?: Array<{
        name: string
        description?: string
        order: number
        machineGroup?: string
        tasks?: Array<{
          ref: number
          name: string
          module: string
          order: number
          params?: string
          when?: string
          hooks?: string
          loop?: string
          timeout?: number
          ignore_errors?: boolean
          retries?: number
          delay?: number
          register?: string
        }>
      }>
    }>
    hooks?: Array<{
      name: string
      module: string
      params?: string
      timeout?: number
      ignore_errors?: boolean
      retries?: number
      delay?: number
    }>
  }>
}

/** 获取方案列表 */
export function getSolutionLibrariesApi(category?: string): Promise<SolutionLibrary[]> {
  return request.get('/solution-libraries', { params: { category } }).then((res) => res.data.data)
}

/** 获取方案详情 */
export function getSolutionLibraryApi(id: number): Promise<SolutionLibrary> {
  return request.get(`/solution-libraries/${id}`).then((res) => res.data.data)
}

/** 创建方案 */
export function createSolutionLibraryApi(data: any): Promise<SolutionLibrary> {
  return request.post('/solution-libraries', data).then((res) => res.data.data)
}

/** 更新方案 */
export function updateSolutionLibraryApi(id: number, data: any): Promise<void> {
  return request.put(`/solution-libraries/${id}`, data).then((res) => res.data)
}

/** 删除方案 */
export function deleteSolutionLibraryApi(id: number): Promise<void> {
  return request.delete(`/solution-libraries/${id}`).then((res) => res.data)
}

/** 导出方案 */
export function exportSolutionLibraryApi(id: number): Promise<OrbitPack> {
  return request.get(`/solution-libraries/${id}/export`).then((res) => res.data.data)
}

// 冲突项
export interface ConflictItem {
  type: string
  name: string
  existing_source: string
}

// 冲突检测响应
export interface ConflictResponse {
  has_conflicts: boolean
  conflicts: ConflictItem[]
  summary: {
    stage_count: number
    variable_count: number
    hook_count: number
    template_count: number
    file_count: number
    workflow_count: number
  }
}

// 导入方案（简单导入，只检查方案名称）
export function importSolutionLibraryApi(data: {
  pack: OrbitPack
}): Promise<SolutionLibrary> {
  return request.post('/solution-libraries/import', data).then((res) => res.data.data)
}

// 应用方案（检测冲突 + 根据决策执行）
export function applySolutionLibraryApi(
  id: number,
  decisions?: Record<string, Record<string, string>>,
  variableValues?: Record<string, string>,
  machineGroupMachines?: Record<string, number[]>
): Promise<ConflictResponse | void> {
  return request.post(`/solution-libraries/${id}/apply`, { decisions, variable_values: variableValues, machine_group_machines: machineGroupMachines }).then((res) => res.data.data)
}

// 文件下载代理
export function downloadFileAPI(url: string): Promise<Blob> {
  return request.post('/solution-libraries/download-proxy', { url }, { responseType: 'blob' }).then((res) => res.data)
}

File: backend/models/workflow/solution_library.go
package workflow

import (
	"time"
)

// SolutionLibrary 方案库（模板市场的基本单位，包含多个模块的组合）
type SolutionLibrary struct {
	ID          uint   `json:"id" gorm:"primaryKey"`
	Name        string `json:"name" gorm:"size:100;not null;uniqueIndex"`
	Description string `json:"description" gorm:"size:500"`
	Category    string `json:"category" gorm:"size:50;index"` // k8s, database, monitoring, etc.
	Version     string `json:"version" gorm:"size:20"`
	Author      string `json:"author" gorm:"size:100"`
	Icon        string `json:"icon" gorm:"size:50"` // mdi icon name
	// 原始数据（导入时存储 OrbitPack JSON，应用后清空）
	PackData string `json:"pack_data" gorm:"type:text"` // OrbitPack JSON
	// 关联ID（JSON数组格式存储，应用后才有值）
	StageIDs    string `json:"stage_ids" gorm:"type:text"`    // [1,2,3]
	VariableIDs string `json:"variable_ids" gorm:"type:text"` // [1,2]
	HookIDs     string `json:"hook_ids" gorm:"type:text"`     // [1]
	TemplateIDs string `json:"template_ids" gorm:"type:text"` // [1,2]
	FileIDs     string `json:"file_ids" gorm:"type:text"`     // [1]
	WorkflowIDs string `json:"workflow_ids" gorm:"type:text"` // [1]
	// 统计字段
	StageCount    int       `json:"stage_count"`
	VariableCount int       `json:"variable_count"`
	HookCount     int       `json:"hook_count"`
	TemplateCount int       `json:"template_count"`
	FileCount     int       `json:"file_count"`
	WorkflowCount int       `json:"workflow_count"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
}

func (SolutionLibrary) TableName() string { return "solution_library" }

// OrbitPack orbit-pack YAML 格式定义
type OrbitPack struct {
	APIVersion string `yaml:"apiVersion" json:"apiVersion"`
	Kind       string `yaml:"kind" json:"kind"`
	Metadata   struct {
		Name        string `yaml:"name" json:"name"`
		Description string `yaml:"description" json:"description"`
		Category    string `yaml:"category" json:"category"`
		Version     string `yaml:"version" json:"version"`
		Author      string `yaml:"author" json:"author"`
	} `yaml:"metadata" json:"metadata"`
	Materials         []PackMaterial         `yaml:"materials,omitempty" json:"materials,omitempty"`
	GlobalVariables   []PackGlobalVariable   `yaml:"global_variables,omitempty" json:"globalVariables,omitempty"`
	Hooks             []PackHook             `yaml:"hooks,omitempty" json:"hooks,omitempty"`
	WorkflowTemplates []PackWorkflowTemplate `yaml:"workflow_templates,omitempty" json:"workflowTemplates,omitempty"`
	MachineGroups     []PackMachineGroup      `yaml:"machine_groups,omitempty" json:"machineGroups,omitempty"`
	Stages            []PackStage            `yaml:"stages,omitempty" json:"stages,omitempty"`
	Workflows         []PackWorkflow         `yaml:"workflows,omitempty" json:"workflows,omitempty"`
}

// PackMachineGroup 打包的机器分组
type PackMachineGroup struct {
	Name        string `yaml:"name" json:"name"`
	Description string `yaml:"description,omitempty" json:"description,omitempty"`
}

// PackMaterial 物料清单
type PackMaterial struct {
	Name        string `yaml:"name" json:"name"`
	Size        int64  `yaml:"size" json:"size"`
	MD5         string `yaml:"md5" json:"md5"`
	DownloadURL string `yaml:"download_url,omitempty" json:"downloadUrl,omitempty"`
}

// PackGlobalVariable 打包的全局变量
type PackGlobalVariable struct {
	Key         string `yaml:"key" json:"key"`
	Type        string `yaml:"type" json:"type"`
	Value       string `yaml:"value" json:"value"`
	Description string `yaml:"description" json:"description"`
	Group       string `yaml:"group" json:"group"`
}

// PackHook 打包的钩子模板
type PackHook struct {
	Name         string `yaml:"name" json:"name"`
	Description  string `yaml:"description,omitempty" json:"description,omitempty"`
	Module       string `yaml:"module" json:"module"`
	Params       string `yaml:"params,omitempty" json:"params,omitempty"`
	Timeout      int    `yaml:"timeout,omitempty" json:"timeout,omitempty"`
	IgnoreErrors bool   `yaml:"ignore_errors,omitempty" json:"ignoreErrors,omitempty"`
	Retries      int    `yaml:"retries,omitempty" json:"retries,omitempty"`
	Delay        int    `yaml:"delay,omitempty" json:"delay,omitempty"`
}

// PackWorkflowTemplate 打包的工作流模板文件
type PackWorkflowTemplate struct {
	Name        string `yaml:"name" json:"name"`
	Description string `yaml:"description,omitempty" json:"description,omitempty"`
	Content     string `yaml:"content" json:"content"`
}

// PackStage 打包的阶段模板
type PackStage struct {
	Name         string     `yaml:"name" json:"name"`
	Description  string     `yaml:"description,omitempty" json:"description,omitempty"`
	MachineGroup string     `yaml:"machine_group,omitempty" json:"machineGroup,omitempty"` // 使用名称而非ID
	Tasks        []PackTask `yaml:"tasks" json:"tasks"`
}

// PackTask 打包的任务
type PackTask struct {
	Ref          int      `yaml:"ref" json:"ref"`
	Name         string   `yaml:"name" json:"name"`
	Module       string   `yaml:"module" json:"module"`
	Order        int      `yaml:"order" json:"order"`
	Params       string   `yaml:"params,omitempty" json:"params,omitempty"`
	When         string   `yaml:"when,omitempty" json:"when,omitempty"`
	Hooks        []string `yaml:"hooks,omitempty" json:"hooks,omitempty"` // 使用名称而非ID
	Loop         string   `yaml:"loop,omitempty" json:"loop,omitempty"`
	Timeout      int      `yaml:"timeout,omitempty" json:"timeout,omitempty"`
	IgnoreErrors bool     `yaml:"ignore_errors,omitempty" json:"ignoreErrors,omitempty"`
	Retries      int      `yaml:"retries,omitempty" json:"retries,omitempty"`
	Delay        int      `yaml:"delay,omitempty" json:"delay,omitempty"`
	Register     string   `yaml:"register,omitempty" json:"register,omitempty"`
}

type PackWorkflow struct {
	Name        string           `yaml:"name" json:"name"`
	Description string           `yaml:"description,omitempty" json:"description,omitempty"`
	StageGroups []PackStageGroup `yaml:"stage_groups,omitempty" json:"stageGroups,omitempty"`
	Hooks       []PackHookGroup  `yaml:"hooks,omitempty" json:"hooks,omitempty"`
}

type PackHookGroup struct {
	Name         string `yaml:"name" json:"name"`
	Module       string `yaml:"module" json:"module"`
	Params       string `yaml:"params,omitempty" json:"params,omitempty"`
	Timeout      int    `yaml:"timeout,omitempty" json:"timeout,omitempty"`
	IgnoreErrors bool   `yaml:"ignore_errors,omitempty" json:"ignoreErrors,omitempty"`
	Retries      int    `yaml:"retries,omitempty" json:"retries,omitempty"`
	Delay        int    `yaml:"delay,omitempty" json:"delay,omitempty"`
}

type PackStageGroup struct {
	Name        string              `yaml:"name" json:"name"`
	Description string              `yaml:"description,omitempty" json:"description,omitempty"`
	Order       int                 `yaml:"order" json:"order"`
	Mode        string              `yaml:"mode" json:"mode"`
	Stages      []PackWorkflowStage `yaml:"stages,omitempty" json:"stages,omitempty"`
}

type PackWorkflowStage struct {
	Name         string             `yaml:"name" json:"name"`
	Description  string             `yaml:"description,omitempty" json:"description,omitempty"`
	Order        int                `yaml:"order" json:"order"`
	MachineGroup string             `yaml:"machine_group,omitempty" json:"machineGroup,omitempty"` // 使用名称而非ID
	Tasks        []PackWorkflowTask `yaml:"tasks" json:"tasks"`
}

type PackWorkflowTask struct {
	Ref          int      `yaml:"ref" json:"ref"`
	Name         string   `yaml:"name" json:"name"`
	Module       string   `yaml:"module" json:"module"`
	Params       string   `yaml:"params,omitempty" json:"params,omitempty"`
	Order        int      `yaml:"order" json:"order"`
	When         string   `yaml:"when,omitempty" json:"when,omitempty"`
	Hooks        []string `yaml:"hooks,omitempty" json:"hooks,omitempty"` // 使用名称而非ID
	Loop         string   `yaml:"loop,omitempty" json:"loop,omitempty"`
	Timeout      int      `yaml:"timeout,omitempty" json:"timeout,omitempty"`
	IgnoreErrors bool     `yaml:"ignore_errors,omitempty" json:"ignoreErrors,omitempty"`
	Retries      int      `yaml:"retries,omitempty" json:"retries,omitempty"`
	Delay        int      `yaml:"delay,omitempty" json:"delay,omitempty"`
	Register     string   `yaml:"register,omitempty" json:"register,omitempty"`
}
